Mathlib/NumberTheory/SiegelsLemma.lean

```diff
@@ -0,0 +1,237 @@
+/-
+Copyright (c) 2024 Fabrizio Barroero. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fabrizio Barroero, Laura Capuano, Amos Turchet
+-/
+import Mathlib.Analysis.Matrix
+import Mathlib.Data.Pi.Interval
+import Mathlib.Tactic.Rify
+
+/-!
+# Siegel's Lemma
+
+In this file we introduce and prove Siegel's Lemma in its most basic version. This is a fundamental
+tool in diophantine approximation and transcendency and says that there exists a "small" integral
+non-zero solution of a non-trivial underdetermined system of linear equations with integer
+coefficients.
+
+## Main results
+
+- `exists_ne_zero_int_vec_norm_le`: Given a non-zero `m × n` matrix `A` with `m < n` the linear
+system it determines has a non-zero integer solution `t` with
+`‖t‖ ≤ ((n * ‖A‖) ^ ((m : ℝ) / (n - m)))`
+
+## Notation
+
+ - `‖_‖ ` : Matrix.seminormedAddCommGroup is the sup norm, the maximum of the absolute values of
+ the entries of the matrix
+
+## References
+
+See [M. Hindry and J. Silverman, Diophantine Geometry: an Introduction][hindrysilverman00].
+-/
+
+/- We set ‖⬝‖ to be Matrix.seminormedAddCommGroup  -/
+attribute [local instance] Matrix.seminormedAddCommGroup
+
+open Matrix Finset
+
+section generalMatrices
+
+variable {m n α : Type*} [Fintype m] [Fintype n] [SeminormedAddCommGroup α]
+
+lemma exists_entry_norm_eq_norm_Matrix [Nonempty m] [Nonempty n] (A : Matrix m n α) :
+    ∃ (i : m) (j : n), ‖A‖ = ‖A i j‖₊ := by
+  simp only [norm_eq_sup_sup_nnnorm]
+  rcases Finset.exists_mem_eq_sup univ univ_nonempty (fun i => univ.sup fun j => ‖A i j‖₊)
+    with ⟨i, hi⟩
+  use i
+  rcases Finset.exists_mem_eq_sup univ univ_nonempty (fun j => ‖A i j‖₊) with ⟨j, hj⟩
+  use j
+  rw [hi.2, hj.2]
+
+end generalMatrices
+
+namespace Int.Matrix
+
+variable (m n : ℕ) (A : Matrix (Fin m) (Fin n) ℤ) (v : Fin n → ℤ) (hn : m < n)
+(hm : 0 < m)
+
+--Some definitions and relative properties
+
+local notation3 "e" => m / ((n : ℝ) - m) --exponent
+local notation3 "B" => Nat.floor (((n : ℝ) * ‖A‖) ^ e)
+-- B' is the vector with all components = B
+local notation3 "B'" => fun _ : Fin n => (B : ℤ)
+-- T is the box [0 B]^n
+local notation3 "T" =>  Finset.Icc 0 B'
+local notation3  "P" => fun i : Fin m => (∑ j : Fin n, B * posPart (A i j))
+local notation3  "N" => fun i : Fin m =>  (∑ j : Fin n, B * (- negPart (A i j)))
+-- S is the box where the image of T goes
+local notation3  "S" => Finset.Icc N P
+
+section preparation
+
+/- In order to apply Pigeohole we need:
+# Step 1: ∀ v ∈  T, (A.mulVec v) ∈  S
+and
+# Step 2: S.card < T.card
+Pigeohole will give different x and y in T with A.mulVec x = A.mulVec y in S
+Their difference is the solution we are looking for
+-/
+
+-- # Step 1: ∀ v ∈  T, (A.mulVec v) ∈  S
+
+private lemma image_T_subset_S : ∀ v ∈ T, (A.mulVec v) ∈ S := by
```

Comment:
```suggestion
private lemma image_T_subset_S : ∀ v ∈ T, A.mulVec v ∈ S := by
```
There are many unnecessary parentheses. I'll flag some of them, but it is best if you try for yourself which ones are really needed.
Also, you can use notation `A *ᵥ v` for the matrix-vector product.